sky/serve/replica_managers.py

```diff
@@ -0,0 +1,804 @@
+        if replica_id in self.down_process_pool:
+            logger.warning(f'Terminate process for replica {replica_id} '
+                           'already exists. Skipping.')
+            return
+
+        def _download_and_stream_logs(info: ReplicaInfo):
+            launch_log_file_name = (
+                serve_utils.generate_replica_launch_log_file_name(
+                    self.service_name, replica_id))
+            local_log_file_name = (
+                serve_utils.generate_replica_local_log_file_name(
+                    self.service_name, replica_id))
+            # Write launch log to local log file
+            with open(local_log_file_name,
+                      'w') as local_file, open(launch_log_file_name,
+                                               'r') as launch_file:
+                local_file.write(launch_file.read())
+            logger.info(f'Syncing down logs for replica {replica_id}...')
+            backend = backends.CloudVmRayBackend()
+            handle = global_user_state.get_handle_from_cluster_name(
+                info.cluster_name)
+            if handle is None:
+                logger.error(f'Cannot find cluster {info.cluster_name} for '
+                             f'replica {replica_id} in the cluster table. '
+                             'Skipping syncing down job logs.')
+                return
+            assert isinstance(handle, backends.CloudVmRayResourceHandle)
+            replica_job_logs_dir = os.path.join(constants.SKY_LOGS_DIRECTORY,
+                                                'replica_jobs')
+            log_file = backend_utils.download_and_stream_latest_job_log(
+                backend,
+                handle,
+                replica_job_logs_dir,
+                log_position_hint='replica cluster',
+                log_finish_hint=f'Replica: {replica_id}')
+            if log_file is not None:
+                with open(local_log_file_name,
+                          'a') as local_file, open(log_file, 'r') as job_file:
+                    local_file.write(job_file.read())
+
+        logger.info(f'Terminating replica {replica_id}...')
+        info = serve_state.get_replica_info_from_id(self.service_name,
+                                                    replica_id)
+        assert info is not None
+
+        if sync_down_logs:
+            _download_and_stream_logs(info)
+
+        logger.info(f'preempted: {info.status_property.preempted}, '
+                    f'replica_id: {replica_id}')
+        log_file_name = serve_utils.generate_replica_down_log_file_name(
+            self.service_name, replica_id)
+        p = multiprocessing.Process(
+            target=serve_utils.RedirectOutputTo(
+                terminate_cluster,
+                log_file_name,
+            ).run,
+            args=(info.cluster_name,),
+        )
+        p.start()
+        self.down_process_pool[replica_id] = p
```

Comment:
The problem is we need to manually poll the `Process` (by `join` or `is_alive`), otherwise, it will never finish. A test script:

```
import multiprocessing, psutil, time

def foo():
    pass

p = multiprocessing.Process(target=foo)
p.start()
time.sleep(1)
print(psutil.Process(p.pid).is_running())
print(psutil.pid_exists(p.pid))
p.join()
```

Output:

```
$ py @temp/temp.py
True
True
```

Reply:
Therefore seems like we need to keep the variable in memory... Do you have any workaround for this problem?

Reply:
Hmm, I see. This is interesting. We can have a TODO for future investigation. It seems we can check if `psutil.Process(p.pid).status() == psutil.STATUS_ZOMBIE` to see if the process has finished.

Reply:
SG. Added a TODO 🫡